### PR TITLE
feat: support search_group_by for milvus(#25324)

### PR DESCRIPTION
**Pull request by MrPresent-Han**
related: #25324

Search GroupBy function, used to aggregate result entities based on a specific scalar column.
several points to mention:

1. Temporarliy, the whole groupby is implemented separated from iterative expr framework **for the first period**
2. In the long term, the groupBy operation will be incorporated into the iterative expr framework:https://github.com/milvus-io/milvus/pull/28166
3. This pr includes some unrelated mocked interface regarding alterIndex due to some unworth-to-mention reasons. All these un-associated content will be removed before the final pr is merged. This version of pr is only for review
4. All other related details were commented in the files comparison

### CODE REVIEW

**Comment by MrPresent-Han** [comment]
rerun ut

File: internal/core/src/query/GroupByOperator.cpp [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+    int64_t topK,
+    Span<T> field_data,
+    std::vector<GroupByValueType>& group_by_values,
+    std::vector<int64_t>& offsets,
+    std::vector<float>& distances,
+    const knowhere::MetricType& metrics_type) {
+    //1.
+    std::unordered_map<T, std::pair<int64_t, float>> groupMap;
+    std::vector<int64_t> tmpOffsets;
+    std::vector<float> tmpDistances;
```
Comment (yah01) [inline internal/core/src/query/GroupByOperator.cpp:166]: pre-allocate memory for the vectors

File: internal/core/src/query/GroupByOperator.cpp [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+    for (auto iter = sortedGroupVals.cbegin(); iter != sortedGroupVals.cend();
+         iter++) {
+        group_by_values.emplace_back(iter->first);
+        offsets.push_back(iter->second.first);
+        distances.push_back(iter->second.second);
+    }
+
+    //4. padding topK results, extra memory consumed will be removed when reducing
+    for (std::size_t idx = groupMap.size(); idx < topK; idx++) {
+        offsets.push_back(INVALID_SEG_OFFSET);
```
Comment (yah01) [inline internal/core/src/query/GroupByOperator.cpp:211]: could we just ignore these? the `INVALID_SEG_OFFSET` makes it impossible to optimize the retrieving with SIMD

File: internal/core/src/query/visitors/ExecPlanNodeVisitor.cpp [hunk 3]
```diff
@@ -123,7 +124,20 @@ ExecPlanNodeVisitor::VectorVisitorImpl(VectorPlanNode& node) {
...
                            search_result);
-
+    if (search_result.iterators.has_value()) {
+        GroupBy(search_result.iterators.value(),
+                node.search_info_,
+                search_result.group_by_values_,
+                *segment,
+                search_result.seg_offsets_,
+                search_result.distances_);
+        AssertInfo(search_result.seg_offsets_.size() ==
```
Comment (yah01) [inline internal/core/src/query/visitors/ExecPlanNodeVisitor.cpp:134]: It's enabled to format string with `AssertInfo/PanicInfo` like:
```
AssertInfo(a == b, "{} != {}", a, b)
```

**Review by yah01 — CHANGES_REQUESTED** [review]

**Review by MrPresent-Han — COMMENTED** [review]
waiting to be reviewed for version1

File: internal/core/src/query/GroupByOperator.cpp [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+    std::unordered_map<T, std::pair<int64_t, float>> groupMap;
+    std::vector<int64_t> tmpOffsets;
+    std::vector<float> tmpDistances;
+
+    //2. do iteration for at most 10 rounds
+    int round = 0;
+    while (round < 10) {
+        int64_t count = 0;
+        while (iterator->HasNext()) {
+            auto nextPair = iterator->Next();
```
Comment (zhengbuqian) [inline internal/core/src/query/GroupByOperator.cpp:173]: auto [offset, dis] = iterator->Next();

File: internal/core/src/index/VectorMemIndex.cpp [hunk 5]
```diff
@@ -521,6 +521,25 @@ VectorMemIndex<T>::Query(const DatasetPtr dataset,
...
     auto num_queries = dataset->GetRows();
     knowhere::Json search_conf = search_info.search_params_;
+    if (search_info.group_by_field_id_.has_value()) {
+        auto result = std::make_unique<SearchResult>();
+        try {
+            knowhere::expected<
+                std::vector<std::shared_ptr<knowhere::IndexNode::iterator>>>
+                iterators_val =
+                    index_.AnnIterator(*dataset, search_conf, bitset);
+            if (iterators_val.has_value()) {
```
Comment (zhengbuqian) [inline internal/core/src/index/VectorMemIndex.cpp:531]: log error if the returned `knowhere::expected` has an non-ok code. I wonder is simply logging the error sufficient? should we wrap/return the error back to the client?

File: internal/core/src/query/GroupByOperator.cpp [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+// limitations under the License.
+#include "GroupByOperator.h"
+#include "common/Consts.h"
+#include "segcore/SegmentSealedImpl.h"
+#include "Utils.h"
+
+namespace milvus {
+namespace query {
+
+knowhere::DataSetPtr
```
Comment (zhengbuqian) [inline internal/core/src/query/GroupByOperator.cpp:24]: remove return value as it is always nullptr

File: internal/core/src/query/GroupByOperator.cpp [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+            float dis = nextPair.second;
+            tmpOffsets.emplace_back(offset);
+            tmpDistances.emplace_back(dis);
+            count++;
+            if (count >= topK) {
+                break;
+            }
+        }
+        round++;
+        GroupOneRound<T>(tmpOffsets, tmpDistances, field_data, groupMap);
```
Comment (zhengbuqian) [inline internal/core/src/query/GroupByOperator.cpp:184]: we don't need the GroupOneRound helper:

```cpp
// 2. do iteration for at most 10 rounds
int max_it_ratio = 10;  // 
int max_iterations = max_it_ratio * topK;
while (max_iterations-- > 0 && iterator->HasNext() && groupMap.size() < topK) {
  auto [offset, dis] = iterator->Next();
  const T& group_by_field_data = field_data[offset];
  auto it = groupMap.find(group_by_field_data);
  if (it == groupMap.end() || it->second->second > dis) {  // or if it->second->second < dis when using IP metric
     groupMap.emplace(std::make_pair(
                group_by_field_data, std::make_pair(offset, dis)));
  }
}
// 3. sort
...
```

File: internal/core/src/query/GroupByOperator.cpp (hunk 2 above)
Comment (MrPresent-Han) [reply internal/core/src/query/GroupByOperator.cpp:211]: Currently, the reduce logic is designed towards n-query, meaning that results for multiple vectors is assembled together into a single vector. So we need to use this TAG to differentiate the 'domain' for each 'qi'. This is actually a placeholder for reduce logic

File: internal/core/src/query/GroupByOperator.cpp (hunk 1 above)
Comment (MrPresent-Han) [reply internal/core/src/query/GroupByOperator.cpp]: changed

File: internal/core/src/query/visitors/ExecPlanNodeVisitor.cpp (hunk 3 above)
Comment (MrPresent-Han) [reply internal/core/src/query/visitors/ExecPlanNodeVisitor.cpp:196]: changed

File: internal/core/src/query/GroupByOperator.cpp (hunk 4 above)
Comment (MrPresent-Han) [reply internal/core/src/query/GroupByOperator.cpp]: changed

File: internal/core/src/index/VectorMemIndex.cpp (hunk 5 above)
Comment (MrPresent-Han) [reply internal/core/src/index/VectorMemIndex.cpp:533]: changed

File: internal/core/src/query/GroupByOperator.cpp (hunk 6 above)
Comment (MrPresent-Han) [reply internal/core/src/query/GroupByOperator.cpp]: changed

File: internal/core/src/query/GroupByOperator.cpp (hunk 7 above)
Comment (MrPresent-Han) [reply internal/core/src/query/GroupByOperator.cpp]: changed

File: internal/core/src/index/VectorMemIndex.cpp [hunk 8]
```diff
@@ -522,6 +522,32 @@ VectorMemIndex<T>::Query(const DatasetPtr dataset,
...
+                    "Returned knowhere iterator has non-ready iterators "
+                    "inside, "
+                    "the group_by result of this segment will be deprecated");
+            }
+        } catch (const std::runtime_error& e) {
+            LOG_ERROR(
+                "Caught error when trying to initialize ann iterators for "
+                "group_by:, "
+                "group_by will not take effect",
+                e.what());
```
Comment (yah01) [inline internal/core/src/index/VectorMemIndex.cpp:545]: need `{}` to indicate where the `e.what()' would be placed in

File: internal/core/src/segcore/Reduce.cpp [hunk 9]
```diff
@@ -90,6 +88,16 @@ ReduceHelper::FilterInvalidSearchResult(SearchResult* search_result) {
     auto segment = static_cast<SegmentInterface*>(search_result->segment_);
     auto& offsets = search_result->seg_offsets_;
     auto& distances = search_result->distances_;
+    bool need_filter_group_by = !search_result->group_by_values_.empty();
+    if (need_filter_group_by) {
+        AssertInfo(search_result->distances_.size() ==
+                       search_result->group_by_values_.size(),
+                   "wrong group_by_values size, size= " +
+                       std::to_string(search_result->group_by_values_.size()) +
```
Comment (yah01) [inline internal/core/src/segcore/Reduce.cpp:96]: these could be formatted

File: internal/core/src/segcore/Reduce.cpp [hunk 10]
```diff
@@ -146,6 +158,16 @@ ReduceHelper::RefreshSearchResult() {
     for (int i = 0; i < num_segments_; i++) {
         std::vector<int64_t> real_topks(total_nq_, 0);
         auto search_result = search_results_[i];
+        bool need_to_handle_group_by = !search_result->group_by_values_.empty();
+        if (need_to_handle_group_by) {
+            AssertInfo(
```
Comment (yah01) [inline internal/core/src/segcore/Reduce.cpp:163]: same as line 93

File: internal/core/src/segcore/Reduce.cpp [hunk 11]
```diff
@@ -417,4 +479,71 @@ ReduceHelper::GetSearchResultDataSlice(int slice_index) {
...
+    std::unique_ptr<milvus::proto::schema::SearchResultData>& search_result,
+    const std::vector<GroupByValueType>& group_by_vals) {
+    auto group_by_field_id = plan_->plan_node_->search_info_.group_by_field_id_;
+    if (group_by_field_id.has_value() && group_by_vals.size() > 0) {
+        auto group_by_values_field =
+            std::make_unique<milvus::proto::schema::ScalarField>();
+        auto group_by_field =
+            plan_->schema_.operator[](group_by_field_id.value());
+        DataType group_by_data_type = group_by_field.get_data_type();
+        for (std::size_t idx = 0; idx < group_by_vals.size(); idx++) {
```
Comment (yah01) [inline internal/core/src/segcore/Reduce.cpp:493]: swap the order of `for` and `switch`, then a batch append would be possible, which could improve the performance

File: internal/proxy/task_search.go [hunk 12]
```diff
@@ -167,15 +170,39 @@ func parseSearchInfo(searchParamsPair []*commonpb.KeyValuePair) (*planpb.QueryIn
...
+	if groupByFieldName != "" {
+		groupByFieldId = -1
+		fields := schema.GetFields()
+		for _, field := range fields {
+			if field.Name == groupByFieldName {
+				groupByFieldId = field.FieldID
+			}
+		}
+		if groupByFieldId == -1 {
+			return nil, 0, fmt.Errorf("GroupByFieldName:%s is invalid, cannot find field_id accordingly",
```
Comment (yah01) [inline internal/proxy/task_search.go:195]: -> `merr.WrapErrFieldNotFound(groupByFieldId)`

File: internal/core/src/segcore/Reduce.cpp [hunk 13]
```diff
@@ -207,7 +236,13 @@ ReduceHelper::ReduceSearchResultForOneNQ(int64_t qi,
 
         pairs_.emplace_back(
             primary_key, distance, search_result, i, offset_beg, offset_end);
```
Comment (zhengbuqian) [inline internal/core/src/segcore/Reduce.cpp:238]: initialize group_by_value_ through SearchResultPair constructor. 

`pairs_.emplace_back(primary_key, distance, search_result, i, offset_beg, offset_end, need_handle_group_by_values ? search_result->group_by_values_[offset_beg] : std::nullopt);`

File: internal/core/src/segcore/Reduce.cpp [hunk 14]
```diff
@@ -388,6 +442,8 @@ ReduceHelper::GetSearchResultDataSlice(int slice_index) {
         // update result topKs
         search_result_data->mutable_topks()->Set(qi - nq_begin, topk_count);
     }
+    //assemble group_by_values
```
Comment (zhengbuqian) [inline internal/core/src/segcore/Reduce.cpp:445]: unnecessary comment

File: internal/core/src/segcore/Reduce.cpp (hunk 10 above)
Comment (MrPresent-Han) [reply internal/core/src/segcore/Reduce.cpp:160]: changed

File: internal/core/src/segcore/Reduce.cpp [hunk 15]
```diff
@@ -417,4 +479,71 @@ ReduceHelper::GetSearchResultDataSlice(int slice_index) {
     return buffer;
 }
 
+void
+ReduceHelper::AssembleGroupByValues(
+    std::unique_ptr<milvus::proto::schema::SearchResultData>& search_result,
+    const std::vector<GroupByValueType>& group_by_vals) {
+    auto group_by_field_id = plan_->plan_node_->search_info_.group_by_field_id_;
+    if (group_by_field_id.has_value() && group_by_vals.size() > 0) {
```
[the remark this answers is not in the document]
Comment (MrPresent-Han) [reply internal/core/src/segcore/Reduce.cpp:475]: changed

File: internal/core/unittest/bench/bench_search.cpp [hunk 16]
```diff
@@ -114,7 +114,7 @@ Search_Sealed(benchmark::State& state) {
     } else if (choice == 1) {
         // ivf
         auto vec = dataset_.get_col<float>(milvus::FieldId(100));
-        auto indexing = GenVecIndexing(N, dim, vec.data());
+        auto indexing = GenVecIndexing(N, dim, vec.data(), knowhere::IndexEnum::INDEX_HNSW);
```
[the remark this answers is not in the document]
Comment (MrPresent-Han) [reply internal/core/unittest/bench/bench_search.cpp:117]: changed

File: pkg/util/typeutil/schema.go [hunk 17]
```diff
@@ -1069,3 +1069,61 @@ func SelectMinPK[T ResultWithID](results []T, cursors []int64) (int, bool) {
...
+				},
+			}
+		}
+		dstScalarField.GetStringData().Data = append(dstScalarField.GetStringData().Data, groupByVal.(string))
+	default:
+		log.Error("Not supported field type from group_by value field", zap.String("field type",
+			srcDataType.String()))
+		return fmt.Errorf("not supported field type from group_by value field: %s",
+			srcDataType.String())
+	}
```
Comment (MrPresent-Han) [inline pkg/util/typeutil/schema.go:1127]: add error to return

**Review by MrPresent-Han — COMMENTED** [review]
changed again on 1228-23

**Comment by MrPresent-Han** [comment]
rerun ut

**Comment by MrPresent-Han** [comment]
rerun ut

**Comment by MrPresent-Han** [comment]
rerun ut

**Comment by MrPresent-Han** [comment]
rerun ut

**Comment by MrPresent-Han** [comment]
rerun ut

**Review by yah01 — COMMENTED** [review]
/lgtm

File: internal/core/src/query/GroupByOperator.cpp [hunk 18]
```diff
@@ -0,0 +1,211 @@
...
+        if (PositivelyRelated(metrics_type))
+            return l > r;
+        return l <= r;
+    };
+    while (iterator->HasNext() && groupMap.size() < topK) {
+        auto [offset, dis] = iterator->Next();
+        const T& row_data = field_data.operator[](offset);
+        auto it = groupMap.find(row_data);
+        if (it == groupMap.end()) {
+            groupMap.insert(
```
Comment (xiaofan-luan) [inline internal/core/src/query/GroupByOperator.cpp:179]: use    groupMap.emplace？

File: internal/core/src/query/GroupByOperator.cpp [hunk 19]
```diff
@@ -0,0 +1,211 @@
...
+
+    //3. sorted based on distances and metrics
+    std::vector<std::pair<T, std::pair<int64_t, float>>> sortedGroupVals(
+        groupMap.begin(), groupMap.end());
+    auto customComparator = [&](const auto& lhs, const auto& rhs) {
+        return dis_closer(lhs.second.second, rhs.second.second);
+    };
+    std::sort(sortedGroupVals.begin(), sortedGroupVals.end(), customComparator);
+
+    //4. save groupBy results
```
Comment (xiaofan-luan) [inline internal/core/src/query/GroupByOperator.cpp:194]: Preserve the result?
    offsets.reserve(topK);
    distances.reserve(topK);

File: internal/core/src/segcore/Reduce.cpp [hunk 20]
```diff
@@ -193,8 +220,14 @@ ReduceHelper::ReduceSearchResultForOneNQ(int64_t qi,
     }
     pk_set_.clear();
     pairs_.clear();
+    group_by_val_set_.clear();
```
Comment (xiaofan-luan) [inline internal/core/src/segcore/Reduce.cpp:223]: there is no reason to keep this map very time?

In fact I would more like to have a separate function to implement groupby but not implement in same functions.

**Pull request merged**